Fix DetailsList onRenderColumnHeaderTooltip not rendering >= 6.18.0

KevinTCoughlin opened the pull request:
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #5703
- [x] Include a change request file using `$ npm run change`

### Description of changes

As noted in #5703, `onRenderColumnHeaderTooltip` has not worked in OUIFR versions `>= 6.18.0`.  This PR fixes that regression by allowing for developers to supply a `onRenderColumnHeaderTooltip` via `onRenderDetailsHeader`. By default, we do not render column header tooltips _by design_.

If the default props are used when providing `renderAs` overrides, then the `IColumn`'s `ariaLabel` will be used as the tooltip's contents on `mouseenter`.

### Focus areas to test

- By default `DetailsList` should not render tooltips for column headers (existing behavior).
- If `onRenderDetailsHeader` is provided, which in-turn passes a render function for the `onRenderColumnHeaderTooltip` prop, then column header tooltips should render on `mouseenter`.

#### Screengrabs

**Before**

![dl_tooltip_basic_before](https://user-images.githubusercontent.com/706967/48103385-af51ce80-e1e3-11e8-8740-583595334ce6.gif)

**After**

![dl_tooltip_basic_after](https://user-images.githubusercontent.com/706967/48103474-17a0b000-e1e4-11e8-9174-37c870d5835c.gif)

#### Testing Locally

Here is a diff that can be applied to the "Basic DetailsList" example which exhibits the above behavior. I've also added unit tests to ensure that the render functions are invoked if provided.

```diff
diff --git a/packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.Basic.Example.tsx b/packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.Basic.Example.tsx
index a4f13c48d..0242d530f 100644
--- a/packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.Basic.Example.tsx
+++ b/packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.Basic.Example.tsx
@@ -3,7 +3,10 @@ import { TextField } from 'office-ui-fabric-react/lib/TextField';
 import { DetailsList, DetailsListLayoutMode, Selection, IColumn, IDetailsList } from 'office-ui-fabric-react/lib/DetailsList';
 import { MarqueeSelection } from 'office-ui-fabric-react/lib/MarqueeSelection';
 import { Checkbox } from 'office-ui-fabric-react/lib/Checkbox';
-import { createRef } from 'office-ui-fabric-react/lib/Utilities';
+import { createRef, IRenderFunction } from 'office-ui-fabric-react/lib/Utilities';
+import { IDetailsHeaderProps } from '../DetailsHeader.types';
+import { DetailsHeader } from '../DetailsHeader';
+import { ITooltipHostProps, TooltipHost } from '../../Tooltip';
 
 const _items: any[] = [];
 
@@ -84,6 +87,7 @@ export class DetailsListBasicExample extends React.Component<
             items={items}
             columns={_columns}
             setKey="set"
+            onRenderDetailsHeader={this._onRenderDetailsHeader}
             layoutMode={DetailsListLayoutMode.fixedColumns}
             selection={this._selection}
             selectionPreservedOnEmptyClick={true}
@@ -103,6 +107,17 @@ export class DetailsListBasicExample extends React.Component<
     }
   }
 
+  private _onRenderDetailsHeader(props: IDetailsHeaderProps, defaultRender?: IRenderFunction<IDetailsHeaderProps>): JSX.Element | null {
+    return (
+      <DetailsHeader
+        {...props}
+        onRenderColumnHeaderTooltip={(props: ITooltipHostProps, defaultRender?: IRenderFunction<ITooltipHostProps>) => (
+          <TooltipHost {...props} />
+        )}
+      />
+    );
+  }
+
   private _getSelectionDetails(): string {
     const selectionCount = this._selection.getSelectedCount();
 ```

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/7013)

KevinTCoughlin commented:
Looks like some tests that were written after this behavior regressed are failing. I'll post an iteration fixing them and explain the changes.

ThomasMichon commented on packages/office-ui-fabric-react/src/components/DetailsList/__snapshots__/DetailsHeader.test.tsx.snap:
```diff
@@ -1474,31 +1510,8 @@ exports[`DetailsHeader renders accessible labels 1`] = `
-        }
-    id="header4-b-tooltip"
-  >
-    Click to sort.
```
I don't see this text emitted anymore. These strings still need to remain as the content of the element pointed to by `aria-describedby`.

KevinTCoughlin replied:
No longer applicable, thank you for the heads-up though.

KevinTCoughlin commented on packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.base.tsx:
```diff
@@ -179,7 +187,7 @@ export class DetailsColumnBase extends BaseComponent<IDetailsColumnProps> {
     defaultRender?: IRenderFunction<ITooltipHostProps>
   ): JSX.Element => {
-    return <span className={tooltipHostProps.hostClassName}>{tooltipHostProps.children}</span>;
+    return <TooltipHost {...tooltipHostProps} />;
```
I've taken another approach given @ThomasMichon's context above, updating this branch forcefully. Thank you Thomas for raising awareness 👍.

KevinTCoughlin commented:
>This PR may be stamping over critical accessibility fixes involving the column header Aria labeling, especially with respect to sort status and button descriptions.

@ThomasMichon, may I request another review. I believe I've addressed the issues you raised regarding default behavior and accessibility regressions. Thank you for bringing those to my attention 👍.
